tensorflow/java/src/gen/cc/op_parser.cc

```diff
@@ -0,0 +1,417 @@
+        // we put the text in bold, otherwise in italic
+        for (; c_iter != text.cend() && *c_iter == '*'; ++count, ++c_iter) {}
+        if (count > 1) {
+          javadoc_text << (emphasis ? "</b>" : "<b>");
+        } else {
+          javadoc_text << (emphasis ? "</i>" : "<i>");
+        }
+        emphasis = !emphasis;
+      } else {
+        javadoc_text << '*';
+      }
+      break;
+    default:
+      javadoc_text << c;
+      break;
+    }
+  }
+  return javadoc_text.str();
+}
+
+}  // namespace
+
+OpParser::OpParser(const OpDef& op_def, const ApiDef& api_def,
+    const string& lib_name, const string& base_package)
+  : op_def_(op_def), op_api_(api_def), lib_name_(lib_name),
+    base_package_(base_package) {
+}
+
+void OpParser::Parse(std::unique_ptr<OpSpec>* op_ptr) {
+  visited_attrs_.clear();
+  next_generic_ = 'T';
+  op_ptr->reset(new OpSpec(op_api_.graph_op_name()));
+  for (const string& next_input_name : op_api_.arg_order()) {
+    for (int i = 0; i < op_def_.input_arg().size(); ++i) {
+      if (op_def_.input_arg(i).name() == next_input_name) {
+        ParseInput(op_def_.input_arg(i), op_api_.in_arg(i), op_ptr->get());
+        break;
+      }
+    }
+  }
+  for (int i = 0; i < op_def_.attr().size(); ++i) {
+    ParseAttribute(op_def_.attr(i), op_api_.attr(i), op_ptr->get());
+  }
+  for (int i = 0; i < op_def_.output_arg().size(); ++i) {
+    ParseOutput(op_def_.output_arg(i), op_api_.out_arg(i), op_ptr->get());
+  }
+  BuildEndpoints(op_ptr->get());
+}
+
+void OpParser::BuildEndpoints(OpSpec* op) {
+  Javadoc op_doc = Javadoc::Create(ParseDocumentation(op_api_.summary()))
+    .details(ParseDocumentation(op_api_.description()));
+  std::vector<Type> op_supertypes;
+  op_supertypes.push_back(Type::Class("PrimitiveOp", "org.tensorflow.op"));
+  std::map<string, const Type*> op_generics;
+  for (const OpSpec::Operand& output : op->outputs()) {
+    // declare generic output parameters at the Op class level
+    const Type& data_type = output.data_type();
+    if (data_type.kind() == Type::GENERIC && !data_type.unknown()
+        && op_generics.find(data_type.name()) == op_generics.end()) {
+      op_generics.insert(std::make_pair(data_type.name(), &data_type));
+      op_doc.add_param_tag("<" + data_type.name() + ">",
+          "data type of output '" + output.var().name() + "'");
+    }
+    // implement the Op as an (iteration of) Operand if it has only one output
+    if (op->outputs().size() == 1) {
+      Type operand_inf(Type::Interface("Operand", "org.tensorflow"));
+      operand_inf.add_parameter(data_type.unknown() ?
+          Type::Class("Object") : data_type);
+      op_supertypes.push_back(output.iterable() ?
+          Type::IterableOf(operand_inf) : operand_inf);
+    }
+  }
+  for (const auto& endpoint_def : op_api_.endpoint()) {
+    std::vector<string> name_tokens = str_util::Split(endpoint_def.name(), ".");
+    // if the endpoint specifies a package, use it, otherwise derive it from the
```

Comment:
Instead of having `lib_name_` be based on the library name, how about we dump all into a single namespace (`org.tensorflow.ops`)? That may also simplify the BUILD rules as we won't need one per op library, we can just generate all operations in one go.

Reply:
You mean all ops in a single Java package?